server/compute/diffexp_cxg.py

```diff
@@ -1,120 +1,65 @@
 import concurrent.futures
-import numpy as np
 
+import numpy as np
 from numba import jit
+from scipy import stats
 
-from server.dataset.cxg_util import pack_selector_from_indices
-from server.common.compute.diffexp_generic import diffexp_ttest_from_mean_var, mean_var_n
+from server.common.constants import XApproximateDistribution
 from server.common.errors import ComputeError
 
-"""
-See the comments in diffexp_generic for a description of this algorithm
-
-This implementation runs directly in-process.  It is multi- threaded, but not particularly scalable.
-Longer term, will likely move to a distributed framework for this.
-
-There are currently no global throttles on simultaneous workers.
-"""
-
-diffexp_thread_executor = None
-max_workers = None
-target_workunit = None
-
 
-def set_config(config_max_workers, config_target_workunit):
-    global max_workers
-    global target_workunit
-    max_workers = config_max_workers
-    target_workunit = config_target_workunit
-
-
-def get_thread_executor():
-    global diffexp_thread_executor
-    if diffexp_thread_executor is None:
-        diffexp_thread_executor = concurrent.futures.ThreadPoolExecutor(max_workers=max_workers)
-    return diffexp_thread_executor
-
-
-def diffexp_ttest(adaptor, maskA, maskB, top_n=8, diffexp_lfc_cutoff=0.01):
-
-    matrix = adaptor.open_array("X")
+def set_config(*args, **kwargs):
+    # TODO: find callers and eliminate calls
+    pass
+
+
+def diffexp_ttest(adaptor, maskA, maskB, top_n=8, diffexp_lfc_cutoff=0.01, arr="X"):
+    """
+    Return differential expression statistics for top N variables.
+
+    Algorithm:
+    - compute fold change
+    - compute Welch's t-test statistic and pvalue (w/ Bonferroni correction)
+    - return top N abs(logfoldchange) where lfc > diffexp_lfc_cutoff
+
+    If there are not N which meet criteria, augment by removing the logfoldchange
+    threshold requirement.
+
+    Notes on alogrithm:
+    - Welch's ttest provides basic statistics test.
+      https://en.wikipedia.org/wiki/Welch%27s_t-test
+    - p-values adjusted with Bonferroni correction.
+      https://en.wikipedia.org/wiki/Bonferroni_correction
+
+    :param adaptor: DataAdaptor instance
+    :param maskA: observation selection mask for set 1
+    :param maskB: observation selection mask for set 2
+    :param top_n: number of variables to return stats for
+    :param diffexp_lfc_cutoff: minimum
+    :param arr: the tdb array to read
+    absolute value returning [ varindex, logfoldchange, pval, pval_adj ] for top N genes
+    :return:  for top N genes, {"positive": for top N genes, [ varindex, foldchange, pval, pval_adj ],
+              "negative": for top N genes, [ varindex, foldchange, pval, pval_adj ]}
+    """
+    matrix = adaptor.open_array(arr)
     row_selector_A = np.where(maskA)[0]
```

Comment:
Using `np.where` is slightly poor form and lacks clarity.  See numpy [docs](https://numpy.org/doc/stable/reference/generated/numpy.where.html):

> When only condition is provided, this function is a shorthand for np.asarray(condition).[nonzero](https://numpy.org/doc/stable/reference/generated/numpy.nonzero.html#numpy.nonzero)(). Using nonzero directly should be preferred, as it behaves correctly for subclasses.

Reply:
Oh, and I apologize if this was originally my error!

Reply:
This was the original code, not sure who wrote it but good point, we should update it!

Reply:
modified to us numpy recommended style